_posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md

```diff
@@ -0,0 +1,201 @@
+    "${FAKEIPA_IMAGE}"
+done
+```
+
+In this setup, we made it so that all the **sushy-tools** containers will listen on the port range running from 8001, 8002,..., while the **Fake IPA** containers have ports 9001, 9002,... This will likely cause a problem if you have more than 1000 **sushy-tools** containers, but if that's the case, simply choose a less crowded range (for e.g. 700x and 900x) will help.
+
+### Step 3: Add the BMH objects
+
+Now that we have **sushy-tools** and **Fake IPA** containers running, we can already generate the manifest for BMH objects, and apply them to the cluster. A BMH object will look like this
+
+```yaml
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {name}-bmc-secret
+  labels:
+      environment.metal3.io: baremetal
+type: Opaque
+data:
+  username: YWRtaW4=
+  password: cGFzc3dvcmQ=
+---
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+  name: {name}
+spec:
+  online: true
+  bmc:
+    address: redfish+http://192.168.222.1:{port}/redfish/v1/Systems/{uuid}
+    credentialsName: {name}-bmc-secret
+  bootMACAddress: {random_mac}
+  bootMode: legacy
+```
+In this manifest:
+- `name` is the node name that we generated in the previous step.
+- `uuid` is the random uuid that we generated for the same node.
+- `random_mac` is a random mac address for the boot. It's NOT the same as the NIC mac address that we generated for the node.
+- `port` is the listening port on one of the **sushy-tools** containers that we created in the previous step. Since every **sushy-tools** and **Fake IPA** container has information
+about ALL the nodes, we can decide what container to locate the "node". In general, it's a good idea to spread them out, so all containers are loaded equally.
+
+We can go ahead and run `kubectl apply -f` on one (or all of) the BMH manifests. What you expect to see is that a BMH object is created, and its state will change from `registering` to `available` after awhile. It means **ironic** acknowledged that the node is valid, in good state and ready to be provisioned.
+
+### Step 4: Deploy the fake nodes to kubernetes clusters
+
+In a normal Metal3 workflow, after having the BMH objects in `available` state, we can go ahead and provision new kubernetes clusters with `clusterctl`. However, with fake nodes, things get a tiny bit more complex. At the end of the provisioning process, **Cluster API** expects that there is a new kubernetes API server created for the new cluster, from which it will check if all nodes are up, all the control planes have `apiserver`, `etcd`, etc. pods up and running, and so on. It is where the [**Fake Kubernetes API Server** (FKAS)]() comes in.
+
+As the **FKAS README** linked above already described how it works, we won't go into details. Basically, we will just need to send **FKAS** a `register` request, and it will give us an IP and a port, which we can plug into our cluster template, and then run `clusterctl generate cluster`.
```

Comment:
I would suggest getting rid of languge like "we won't , we will just ..."   It is just my personal opinion but usually in technical writings this personal tone is improper. But I know in blog posts even in technical topics folks use it. 

Up to you, I would not use it.

Reply:
Thanks Adam. That's a good advice, the reason I used a rather casual language was to make this post a bit friendlier to newcomers (Same reason why I went on and on about explaining how Metal3, ironic, etc. work), but yeah, I agree that those words aren't always needed.